### PR TITLE
Separate winsock.h inclusion, add more direct includes

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
This PR improves Windows compilation time.

One more step to help fixing #14055

Kind of Follow-up of an old PR #4188 :
> Replace usage of  `struct timeval` by struct OPENSSL_timeval.

I measure a small compile time improvement : around 1 minute less per build.
Lets compare AppVeyor timing.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: e_winsock.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_E_WINSOCK_H
+# define OSSL_E_WINSOCK_H
+
+#  ifdef WINDOWS
+#   if !defined(_WIN32_WCE) && !defined(_WIN32_WINNT)
+       /*
+        * Defining _WIN32_WINNT here in e_os.h implies certain "discipline."
```
Comment (paulidale) [inline e_winsock.h:16]: It's not `e_os.h` anymore.

File: ssl/ssl_local.h [hunk 2]
```diff
@@ -1889,6 +1888,16 @@ pitem *pqueue_iterator(pqueue *pq);
 pitem *pqueue_next(piterator *iter);
 size_t pqueue_size(pqueue *pq);
 
+/* Forward-declare struct timeval. On Windows, it is defined in winsock2.h and
+ * Windows headers define too many macros to be included in public headers.
+ * However, only a forward declaration is needed. */
+struct timeval;
+
+struct OPENSSL_timeval {
```
Comment (paulidale) [inline ssl/ssl_local.h:1896]: I wonder why we're duplicating a system structure.  Would it suffice leave just the forward declaration of `struct timeval` and insist that callers who want to get inside include `winsock2.h`?

I'm concerned about confusion as to which structure should be used where.

Comment (FdaSilvaYY) [reply ssl/ssl_local.h:1896]: I replace one`struct OPENSSL_timeval` member by this one, 20 line below
to be able to cut the #include dependency to `winsock.h`

Else, I will have to add it to every dtls source file in this folder.

See windows definition of [`struct timeval`](https://docs.microsoft.com/en-us/windows/win32/api/winsock/ns-winsock-timeval)

My quick reading of #4188 make me understand that you hit this issue.

Comment (FdaSilvaYY) [reply ssl/ssl_local.h:1896]: Oups, with this change, break the call to BIO_CTRL_DGRAM_SET_NEXT_TIMEOUT,
who is expecting a pointer to a 'struct timeval' object.

**Review by FdaSilvaYY — COMMENTED** [review]
Back to WIP

**Comment by paulidale** [comment]
That's exactly the kind of problem I was worried about :)

File: ssl/ssl_local.h (hunk 2 above)
Comment (FdaSilvaYY) [reply ssl/ssl_local.h]: I try to fix this in the simplest way.
There no need to store a true `timeval` struct as member of `DTLS1_STATE`

**Comment by FdaSilvaYY** [comment]
Ping @openssl Ready for review.
I found a way to minimize my changes.

**Review by paulidale — APPROVED** [review]
The timeval changes seem a little out of place.

**Comment by DDvO** [comment]
Nice that you are handling a further of those nasty issues incurred by Winsock.

I just noticed that you need to add `#include "e_os.h"` to many files, which is cumbersome.
Unfortunately, commit 62e0bc4058c4538678679fe793c6549a0ab86275, which introduces a central `"internal/common.h"` that already contains `#include "e_os.h"` has not been merged yet because it is part of https://github.com/openssl/openssl/pull/15847 which has been postponed to post-beta.
I suggest cherry-picking that commit here to simplify matters already now.

**Comment by t8m** [comment]
Unfortunately this needs another rebase.

File: apps/ecparam.c [hunk 3]
```diff
@@ -9,6 +9,8 @@
  */
 
 #include <string.h>
+#include "internal/e_os.h" /* for strcmp */
+
```
Comment (t8m) [inline apps/ecparam.c:13]: This is probably not needed anymore

File: ssl/d1_lib.c [hunk 4]
```diff
@@ -259,8 +261,23 @@ long dtls1_ctrl(SSL *ssl, int cmd, long larg, void *parg)
     return ret;
 }
 
+<<<<<<< HEAD
```
Comment (beldmit) [inline ssl/d1_lib.c:264]: Looks weird

Comment (FdaSilvaYY) [reply ssl/d1_lib.c]: fixed, rebasing error !

**Comment by FdaSilvaYY** [comment]
I fully fix the last compilation and rebase issues.
should be good for review.

File: apps/ecparam.c (hunk 3 above)
Comment (FdaSilvaYY) [reply apps/ecparam.c]: I push a fixup.

**Comment by t8m** [comment]
The CI failures are relevant.

File: crypto/evp/ec_support.c [hunk 5]
```diff
@@ -7,7 +7,9 @@
  * https://www.openssl.org/source/license.html
  */
 
+
```
Comment (t8m) [inline crypto/evp/ec_support.c:10]: Please drop this extra empty line

File: ssl/ssl_lib.c [hunk 6]
```diff
@@ -9,9 +9,11 @@
  * https://www.openssl.org/source/license.html
  */
 
-#include <stdio.h>
-#include "ssl_local.h"
 #include "internal/e_os.h"
+
```
Comment (t8m) [inline ssl/ssl_lib.c:13]: Why this empty line

**Comment by FdaSilvaYY** [comment]
I push some changes as a spin-off in #20602 .
Will rebase after its merge,so no need to review this one quickly.

**Comment by t8m** [comment]
CI is relevant.

**Comment by t8m** [comment]
Hmm... we should not have sprintf() calls in the library - they should be replaced with BIO_sprintf().

**Comment by FdaSilvaYY** [comment]
> CI is relevant.

I push a quick fixup using a _CodeSpaces_ ...

**Comment by t8m** [comment]
The style check is relevant.

File: include/internal/e_winsock.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+       */
+#   include <winsock2.h>
+#   include <ws2tcpip.h>
+      /*
+       * Clang-based C++Builder 10.3.3 toolchains cannot find C inline
+       * definitions at link-time.  This header defines WspiapiLoad() as an
+       * __inline function.  https://quality.embarcadero.com/browse/RSP-33806
+       */
+#   if !defined(__BORLANDC__) || !defined(__clang__)
+#      include <wspiapi.h>
```
Comment (t8m) [inline include/internal/e_winsock.h:46]: Nit: too much indentation here

Comment (FdaSilvaYY) [reply include/internal/e_winsock.h:46]: Yes, one last nit, thanks.
Will push it tonight (CEST).

File: include/internal/e_winsock.h [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+       */
+#   include <winsock2.h>
+#   include <ws2tcpip.h>
+      /*
+       * Clang-based C++Builder 10.3.3 toolchains cannot find C inline
+       * definitions at link-time.  This header defines WspiapiLoad() as an
+       * __inline function.  https://quality.embarcadero.com/browse/RSP-33806
+       */
+#   if !defined(__BORLANDC__) || !defined(__clang__)
+#       include <wspiapi.h>
```
Comment (t8m) [inline include/internal/e_winsock.h:46]: Still too much indent here:
```
#    include <wspiapi.h>
```

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**